Windows build failure on CUDA 11.2

This fails with CuPy v8.6.0 and likely v9.0.0rc1 as well: https://github.com/conda-forge/cupy-feedstock/pull/113#issuecomment-807202827
```
    nvcc warning : The 'compute_35', 'compute_37', 'compute_50', 'sm_35', 'sm_37' and 'sm_50' architectures are deprecated, and may be removed in a future release (Use -Wno-deprecated-gpu-targets to suppress warning).
    cupy_thrust.cu
    cupy/cuda/cupy_thrust.cu(145): error C3615: constexpr function 'thrust::less<thrust::complex<float>>::operator ()' cannot result in a constant expression
    cupy/cuda/cupy_thrust.cu(148): note: failure was caused by call of undefined function or one not declared 'constexpr'
    cupy/cuda/cupy_thrust.cu(148): note: see usage of '_cmp_less'
    cupy/cuda/cupy_thrust.cu(157): error C3615: constexpr function 'thrust::less<thrust::complex<double>>::operator ()' cannot result in a constant expression
    cupy/cuda/cupy_thrust.cu(160): note: failure was caused by call of undefined function or one not declared 'constexpr'
    cupy/cuda/cupy_thrust.cu(160): note: see usage of '_cmp_less'
    cupy/cuda/cupy_thrust.cu(169): error C3615: constexpr function 'thrust::less<thrust::tuple<unsigned __int64,thrust::complex<float>,thrust::null_type,thrust::null_type,thrust::null_type,thrust::null_type,thrust::null_type,thrust::null_type,thrust::null_type,thrust::null_type>>::operator ()' cannot result in a constant expression
    cupy/cuda/cupy_thrust.cu(172): note: failure was caused by call of undefined function or one not declared 'constexpr'
    cupy/cuda/cupy_thrust.cu(172): note: see usage of '_tuple_less'
    cupy/cuda/cupy_thrust.cu(181): error C3615: constexpr function 'thrust::less<thrust::tuple<unsigned __int64,thrust::complex<double>,thrust::null_type,thrust::null_type,thrust::null_type,thrust::null_type,thrust::null_type,thrust::null_type,thrust::null_type,thrust::null_type>>::operator ()' cannot result in a constant expression
    cupy/cuda/cupy_thrust.cu(184): note: failure was caused by call of undefined function or one not declared 'constexpr'
    cupy/cuda/cupy_thrust.cu(184): note: see usage of '_tuple_less'
    cupy/cuda/cupy_thrust.cu(223): error C3615: constexpr function 'thrust::less<float>::operator ()' cannot result in a constant expression
    cupy/cuda/cupy_thrust.cu(226): note: failure was caused by call of undefined function or one not declared 'constexpr'
    cupy/cuda/cupy_thrust.cu(226): note: see usage of '_real_less'
    cupy/cuda/cupy_thrust.cu(235): error C3615: constexpr function 'thrust::less<double>::operator ()' cannot result in a constant expression
    cupy/cuda/cupy_thrust.cu(238): note: failure was caused by call of undefined function or one not declared 'constexpr'
    cupy/cuda/cupy_thrust.cu(238): note: see usage of '_real_less'
    cupy/cuda/cupy_thrust.cu(247): error C3615: constexpr function 'thrust::less<thrust::tuple<unsigned __int64,float,thrust::null_type,thrust::null_type,thrust::null_type,thrust::null_type,thrust::null_type,thrust::null_type,thrust::null_type,thrust::null_type>>::operator ()' cannot result in a constant expression
    cupy/cuda/cupy_thrust.cu(250): note: failure was caused by call of undefined function or one not declared 'constexpr'
    cupy/cuda/cupy_thrust.cu(250): note: see usage of '_tuple_less'
    cupy/cuda/cupy_thrust.cu(259): error C3615: constexpr function 'thrust::less<thrust::tuple<unsigned __int64,double,thrust::null_type,thrust::null_type,thrust::null_type,thrust::null_type,thrust::null_type,thrust::null_type,thrust::null_type,thrust::null_type>>::operator ()' cannot result in a constant expression
    cupy/cuda/cupy_thrust.cu(262): note: failure was caused by call of undefined function or one not declared 'constexpr'
    cupy/cuda/cupy_thrust.cu(262): note: see usage of '_tuple_less'
    cupy/cuda/cupy_thrust.cu(287): error C3615: constexpr function 'thrust::less<__half>::operator ()' cannot result in a constant expression
    cupy/cuda/cupy_thrust.cu(288): note: failure was caused by call of undefined function or one not declared 'constexpr'
    cupy/cuda/cupy_thrust.cu(288): note: see usage of '_real_less'
    cupy/cuda/cupy_thrust.cu(297): error C3615: constexpr function 'thrust::less<thrust::tuple<unsigned __int64,__half,thrust::null_type,thrust::null_type,thrust::null_type,thrust::null_type,thrust::null_type,thrust::null_type,thrust::null_type,thrust::null_type>>::operator ()' cannot result in a constant expression
    cupy/cuda/cupy_thrust.cu(300): note: failure was caused by call of undefined function or one not declared 'constexpr'
    cupy/cuda/cupy_thrust.cu(300): note: see usage of '_tuple_less'
    error: command 'D:\\bld\\cupy_1616682886118\\_build_env\\Library\\bin\\nvcc.bat' failed with exit status 2
```
My theory is that on our Windows CI we use VS2015, but on Conda-Forge we use VS2017. According to the VS docs the C3615 error is only raised with VS2017, so we didn't catch it.

--- a/cupy/cuda/cupy_thrust.cu
+++ b/cupy/cuda/cupy_thrust.cu
@@ -6,7 +6,12 @@
 #include <thrust/sort.h>
 #include <thrust/tuple.h>
 #include <thrust/execution_policy.h>
+#if (__CUDACC_VER_MAJOR__ >11 || (__CUDACC_VER_MAJOR__ == 11 && __CUDACC_VER_MINOR__ >= 2))
+// This is used to avoid a problem with constexpr in functions declarations introduced in
+// cuda 11.2, MSVC 15 does not fully support it so we need a dummy constexpr declaration
+// that is provided by this header. However optional.h is not present in old CUDA versions.
 #include <thrust/optional.h>
+#endif
 #include "cupy_thrust.h"
 
 

--- a/cupy/cuda/cupy_thrust.cu
+++ b/cupy/cuda/cupy_thrust.cu
@@ -9,7 +9,8 @@
 #if (__CUDACC_VER_MAJOR__ >11 || (__CUDACC_VER_MAJOR__ == 11 && __CUDACC_VER_MINOR__ >= 2))
 // This is used to avoid a problem with constexpr in functions declarations introduced in
 // cuda 11.2, MSVC 15 does not fully support it so we need a dummy constexpr declaration
-// that is provided by this header. However optional.h is not present in old CUDA versions.
+// that is provided by this header. However optional.h is only available
+// starting CUDA 10.1
 #include <thrust/optional.h>
 #endif
 #include "cupy_thrust.h"

--- a/cupy/cuda/cupy_thrust.cu
+++ b/cupy/cuda/cupy_thrust.cu
@@ -6,6 +6,7 @@
 #include <thrust/sort.h>
 #include <thrust/tuple.h>
 #include <thrust/execution_policy.h>
+#include <thrust/optional.h>
 #include "cupy_thrust.h"
 
 
@@ -61,9 +62,7 @@ public:
 template <typename T>
 __host__ __device__ __forceinline__ 
 #if (__CUDACC_VER_MAJOR__ >11 || (__CUDACC_VER_MAJOR__ == 11 && __CUDACC_VER_MINOR__ >= 2))
-#ifndef _WIN32
-constexpr
-#endif
+THRUST_OPTIONAL_CPP11_CONSTEXPR
 #endif
 bool _tuple_less(const tuple<size_t, T>& lhs,
                                                      const tuple<size_t, T>& rhs) {
@@ -95,9 +94,7 @@ bool _tuple_less(const tuple<size_t, T>& lhs,
 template <typename T>
 __host__ __device__ __forceinline__
 #if (__CUDACC_VER_MAJOR__ >11 || (__CUDACC_VER_MAJOR__ == 11 && __CUDACC_VER_MINOR__ >= 2))
-#ifndef _WIN32
-constexpr
-#endif
+THRUST_OPTIONAL_CPP11_CONSTEXPR
 #endif
 bool _cmp_less(const T& lhs, const T& rhs) {
     bool lhsRe = isnan(lhs.real());
@@ -192,9 +189,7 @@ bool less< tuple<size_t, complex<double>> >::operator() (
 template <typename T>
 __host__ __device__ __forceinline__
 #if (__CUDACC_VER_MAJOR__ >11 || (__CUDACC_VER_MAJOR__ == 11 && __CUDACC_VER_MINOR__ >= 2))
-#ifndef _WIN32
-constexpr
-#endif
+THRUST_OPTIONAL_CPP11_CONSTEXPR
 #endif
 bool _real_less(const T& lhs, const T& rhs) {
     #if  (defined(__CUDA_ARCH__) || defined(__HIP_DEVICE_COMPILE__))

